kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -598,79 +598,96 @@ std::optional<uint64_t> KvbAppFilter::getNextEventGroupId(std::shared_ptr<EventG
-    } else {
-      std::stringstream msg;
-      msg << "No more event groups in storage";
-      LOG_WARN(logger_, msg.str());
-      break;
-    }
-    eg_data_state_->curr_trid_event_group_id++;
-    LOG_DEBUG(logger_,
-              "Global_event_group_id: " << global_event_group_id
-                                        << ", trid_event_group_id: " << eg_data_state_->curr_trid_event_group_id
-                                        << ", event_group_id_start: " << event_group_id_start);
-    // we are not at the starting point in the event group list yet, let's keep incrementing
-    // eg_data_state_->curr_trid_event_group_id until we reach the start
-    if (eg_data_state_->curr_trid_event_group_id < event_group_id_start) {
-      continue;
-    }
-    auto event_group = getEventGroup(global_event_group_id);
+  while (ext_eg_id <= newest_external_eg_id) {
+    // Get events and filter
+    auto event_group = getEventGroup(global_eg_id);
     if (event_group.events.empty()) {
       std::stringstream msg;
-      msg << "EventGroup doesn't exist for valid event_group_id: " << global_event_group_id;
+      msg << "EventGroup empty/doesn't exist for global event group " << global_eg_id;
       throw KvbReadError(msg.str());
     }
-    KvbFilteredEventGroupUpdate update{eg_data_state_->curr_trid_event_group_id,
-                                       filterEventsInEventGroup(eg_data_state_->curr_trid_event_group_id, event_group)};
+    KvbFilteredEventGroupUpdate update{ext_eg_id, filterEventsInEventGroup(global_eg_id, event_group)};
+
+    // Push to queue
     while (!stop_execution) {
-      if (queue_out.push(update)) {
-        break;
+      if (queue_out.push(update)) break;
+    }
+    if (stop_execution) break;
+    if (ext_eg_id == newest_external_eg_id) break;
+
+    // Update next public or private ids; Only one needs to be udpated
+    if (is_previous_private) {
+      if (next_pvt_eg_id > newest_private_eg_id) {
+        pvt_global_id = std::numeric_limits<uint64_t>::max();
+      } else {
+        std::tie(pvt_global_id, pvt_external_id) = getValueFromTagTable(next_pvt_key);
+      }
+    } else {
+      if (next_pub_eg_id > newest_public_eg_id) {
+        pub_global_id = std::numeric_limits<uint64_t>::max();
+      } else {
+        std::tie(pub_global_id, std::ignore) = getValueFromTagTable(next_pub_key);
       }
     }
-    if (stop_execution) {
-      break;
+    // The lesser global event group id is the next update for the client
+    if (pvt_global_id < pub_global_id) {
+      global_eg_id = pvt_global_id;
+      is_previous_private = true;
+      ConcordAssertEQ(ext_eg_id + 1, pvt_external_id);
+      next_pvt_key = client_id_ + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(++next_pvt_eg_id);
+    } else {
+      global_eg_id = pub_global_id;
+      is_previous_private = false;
+      next_pub_key = kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(++next_pub_eg_id);
```

Comment:
If there is a scenario wherein `pvt_global_id` == `pub_global_id` == INT_MAX, i.e., none of the next updates exist yet in storage, we will still enter this else block and increment `next_pub_eg_id`? Can we skip an update here, or am I missing something?

Reply:
You're right, we can skip this whole if-else block and `break` early if both are MAX. Thank you!